### PR TITLE
Add `apply_block_events` and `apply_block_connected_to_events`

### PR DESCRIPTION
**Pull request by tnull**
### Description

Previously, we added a new `Wallet::apply_update_events` method that returned `WalletEvent`s. Unfortunately, no corresponding APIs were added for the `apply_block` counterparts. Here we fix this omission.

### Notes to the reviewers

I opened this towards the `release-2.2` branch, but it would probably need another release branch. Or let me know if you prefer to open it against master (which seems to be lacking `apply_update_events` currently though).

I also added no test coverage given that none seems to exist for `Wallet::apply_block` in the first place. Let me know if I should add something here.

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `just p` before pushing

#### New Features:

* [ ] I've added tests for the new feature
* [x] I've added docs for the new feature

cc @notmandatory

### CODE REVIEW

File: wallet/src/wallet/mod.rs [hunk 1]
```diff
@@ -2560,6 +2560,41 @@ impl Wallet {
...
+    /// See [`apply_update_events`] for more information on the returned [`WalletEvent`]s.
+    ///
+    /// [`apply_block_connected_to_events`]: Self::apply_block_connected_to_events
+    /// [`apply_update_events`]: Self::apply_update_events
+    pub fn apply_block_events(
+        &mut self,
+        block: &Block,
+        height: u32,
+    ) -> Result<Vec<WalletEvent>, CannotConnectError> {
+        let connected_to = match height.checked_sub(1) {
```
[the remark this answers is not in the document]
Comment (ValuedMammal) [reply wallet/src/wallet/mod.rs]: You could have `apply_block` call the new `apply_block_events` and map the return value to `()`.

[the remark this answers is not in the document]
Comment (notmandatory) [reply wallet/src/wallet/mod.rs]: In #319 (in v3.0) my plan is to always return events. But for this PR I agree it's best keep the two paths separate so events are only calculated on the `_events` functions.

File: wallet/src/wallet/mod.rs [hunk 2]
```diff
@@ -2560,6 +2560,41 @@ impl Wallet {
...
+                hash: block.block_hash(),
+            },
+        };
+        self.apply_block_connected_to_events(block, height, connected_to)
+            .map_err(|err| match err {
+                ApplyHeaderError::InconsistentBlocks => {
+                    unreachable!("connected_to is derived from the block so must be consistent")
+                }
+                ApplyHeaderError::CannotConnect(err) => err,
+            })
```
[the remark this answers is not in the document]
Comment (tnull) [reply wallet/src/wallet/mod.rs]: Alright, done!

**Comment by notmandatory** [comment]
I added a few tests for `apply_block_events` based on the relevant (non-mempool related) tests I made for `apply_update_events`.  I think these are the only ones needed for applying on-chain blocks, but free to modify these or add more.

**Comment by ValuedMammal** [comment]
> For the branching question. I'd prefer to have a `release/2.x` branch that contains a tag for each published release. I don't think we'll need branches to do bug fixes separately for 2.1,2.2 etc.

That's a good suggestion.
